libs/intro/generate/generate-thumbnails.ts

```diff
@@ -8,100 +8,138 @@ import {
+  switchMap(([objectIds, gSlides]: [string[], GSlides]) =>
+    forkJoin(
+      objectIds.map(
+        (pageObjectId: string): ObservableInput<string> =>
+          from(
+            gSlides.presentations.pages.getThumbnail({
+              presentationId: CODELAB_PRESENTATION_ID,
+              pageObjectId
+            })
+          ).pipe(map(response => response.data.contentUrl))
+      ) as ObservableInput<string>[]
+    )
+  ),
+  tap((): void => {
+    if (!existsSync(ASSETS_SLIDES_PATH)) {
+      mkdirSync(ASSETS_SLIDES_PATH);
+    }
+  }),
+  tap(contentUrls => {
+    contentUrls.forEach((contentUrl, i) => {
+      const dest = resolve(ASSETS_SLIDES_PATH, `slide-${i}.png`);
+      get(contentUrl, response => {
+        response.pipe(createWriteStream(dest));
       });
+    });
+  }),
+  switchMap(contentUrls =>
+    from(
+      outputJSON(
+        SLIDES_METADATA_PATH,
+        { count: contentUrls.length, contentUrls },
+        { spaces: 2 }
+      )
+    )
+  )
+);
 
-      return response.data.contentUrl;
+const getNewToken$ = oAuth2Client$.pipe(
+  tap(oAuth2Client => {
+    const authUrl = oAuth2Client.generateAuthUrl({
+      access_type: 'offline',
+      scope: ['https://www.googleapis.com/auth/presentations.readonly']
+    });
+    console.log('Authorize this app by visiting this url:', authUrl);
+  }),
+  map(() =>
+    createInterface({
+      input: process.stdin,
+      output: process.stdout
     })
-  );
+  ),
+  switchMap(
+    (readline): Observable<string> => {
+      return new Observable(observer =>
+        readline.question('Enter the code from that page here: ', code => {
+          readline.close();
+          observer.next(code);
+        })
+      );
+    }
+  ),
+  withLatestFrom(oAuth2Client$),
+  switchMap(([code, oAuth2Client]) => {
+    return new Observable(observer =>
+      oAuth2Client.getToken(code, (err, token) => {
+        if (err) {
+          observer.error(err);
+          return console.error('Error retrieving access token', err);
+        }
 
-  if (!existsSync(ASSETS_SLIDES_PATH)) {
-    mkdirSync(ASSETS_SLIDES_PATH);
-  }
+        oAuth2Client.setCredentials(token);
+        observer.next(token);
+      })
+    );
+  }),
+  switchMap(token => from(outputJSON(TOKEN_PATH, token, { spaces: 2 })))
+);
 
-  for (let i = 0; i < contentUrls.length; i += 1) {
-    const dest = resolve(ASSETS_SLIDES_PATH, `slide-${i}.png`);
-    get(contentUrls[i], response => {
-      response.pipe(createWriteStream(dest));
-    });
-  }
+const token$ = from<ObservableInput<Token>>(readJSON(TOKEN_PATH)).pipe(
+  withLatestFrom(oAuth2Client$),
+  tap(([token, oAuth2Client]) => {
+    oAuth2Client.setCredentials(token);
+  }),
+  catchError(() => getNewToken$),
+  switchMap(() => slides$),
+  finalize((): void => console.log('Completed ✅ '))
+);
 
-  await outputJSON(
-    SLIDES_METADATA_PATH,
-    { count: slides.length, contentUrls },
-    { spaces: 2 }
-  );
-}
+token$.pipe(take(1)).subscribe(() => {});
```

Comment:
While I think overall using rxjs improves things here, I find the structure a bit confusing:

Lots of work happens as a side effects (in tap)
And the actual work happens by subscribing to a token$ observable? Why?